### PR TITLE
Implement clj:str

### PR DESCRIPTION
**Pull request by skovsgaard**
This is an attempt at implementing `clj:str/?`, which constructs a string from an arbitrary number of values.

Please poke holes in this solution as I'm sure to have overlooked something.

### CODE REVIEW

File: test/clj-tests.lfe [hunk 1]
```diff
@@ -693,3 +693,11 @@
...
   (is-match 2 (clj:dec 3))
   (is-match 4.0 (clj:dec 5.0)))
+
+(deftest str
+  (is-equal () (clj:str))
+  (is-equal "123" (clj:str 1 2 3))
+  (is-equal "abc" (clj:str "a" "b" "c"))
+  (is-equal "abc" (clj:str 'a 'b "c"))
+  (is-equal "1a2b" (clj:str 1 'a 2 "b"))
+  (is-equal "2.00c" (clj:str 2.0 'c)))
```
Comment (yurrriq) [inline test/clj-tests.lfe]: Can we add a test or two with a `let` and confirm that `str`-ing a variable works as desired too?

Also, for cases like this where the expected values are literals, I think it's better to use `is-match` so we get more descriptive errors.

**Review by yurrriq — CHANGES_REQUESTED** [review]

File: test/clj-tests.lfe (hunk 1 above)
Comment (skovsgaard) [reply test/clj-tests.lfe]: Yup. I'll make those changes as soon as I can. It should be later today.

**Comment by yurrriq** [comment]
It basically does what your `clj:str/1` does but for a single term instead of a list of them. It works fine for me :confused: but notably handles individual characters as integers, which should be documented if not avoided.

``` lfe
;; faked based on memory, but you get the idea
(= "65" (clj:str #\a)) 
```

**Comment by skovsgaard** [comment]
I'm not entirely sure how to avoid that happening. Characters are just integers and I'm not sure how I'd go about differentiating when someone wants a string of "65" and when they want an upper case A.

Where should I document this if it cannot be avoided?

**Comment by yurrriq** [comment]
In `doc/src/lfe_clj.3.md` or similar. Then you can regenerate the docs (with `pandoc` installed) via `make docs` or similar. I'm mobile now.

**Comment by skovsgaard** [comment]
Okay. I'll try doing that. Thank you.

**Comment by yurrriq** [comment]
Oh, I meant something like `(let ((x ...)) (clj:str x ...))`.

**Comment by skovsgaard** [comment]
Okay, I added the docs and generated again. I really hope there's a project policy to only use `1.17.1` of Pandoc. Otherwise I just jumped through a lot of unnecessary install hoops.

Also should I replace the `let` tests with what you meant or just add it after them?

**Comment by yurrriq** [comment]
The more tests the merrier. We don't have a strict version requirement for `pandoc` but it can't hurt to use that same as Robert. I think I use a more recent version...

**Comment by skovsgaard** [comment]
Cool. I added two tests for use within a `let`.

**Comment by yurrriq** [comment]
Comparing the Erlang implementation of `lfe_io_write:term/1` (which is what `lfe_io:print1/1` calls under the hood) and the current`clj:str` macro. I think we should replace the ad hoc `match-lambda` with a call to `lfe_io:print1/1`. What was the issue you had with that, @skovsgaard?

``` erlang
term(S) -> term(S, -1).                         %All the way

term(_, 0) -> "...";
term(Symb, _) when is_atom(Symb) -> symbol(Symb);
term(Numb,_ ) when is_integer(Numb) -> integer_to_list(Numb);
term(Numb, _) when is_float(Numb) -> io_lib_format:fwrite_g(Numb);
term(List, D) when is_list(List) ->
    [$(,list(List, D-1),$)];
term({}, _) -> "#()";
term(Vec, D) when is_tuple(Vec) ->
    Es = tuple_to_list(Vec),
    ["#(",list(Es, D-1),")"];
term(Bit, _) when is_bitstring(Bit) ->
    bitstring(Bit);
term(Map, D) when ?IS_MAP(Map) -> map(Map, D);
term(Other, D) ->                               %Use standard Erlang for rest
    io_lib:write(Other, D).
```

``` lfe
(defmacro str args
  "Construct a string from an arbitrary number of scalar values."
  `(lists:flatmap
    (match-lambda
      ([x] (when (is_integer x))
       (integer_to_list x))
      ([x] (when (is_float x))
       (float_to_list x '(#(decimals 2))))
      ([x] (when (is_atom x))
       (atom_to_list x))
      ([x] (when (is_binary x))
       (binary_to_list x))
      ([x] x))
    (list ,@args)))
```

I guess we'll want an `if` form to check for `clj:string?` so we don't print `"string"` as `(115 116 114 105 110 103)`.

**Comment by yurrriq** [comment]
``` lfe
> (include-lib "lfe/include/clj.lfe")
identical?
> (defmacro str args
    `(lists:flatmap
       (fn [arg] (cond-> arg (not (clj:string? arg)) (lfe_io:print1)))
       (list ,@args)))
str
> (let ((test "test")) (str test 'ing " " 123))
"testing 123"
```

**Comment by skovsgaard** [comment]
I didn't actually have an issue with it. I was just curious about the benefit of it, but if it reduces the amount of code and/or gets rid of the issue with characters, I'm all for it.

**Comment by skovsgaard** [comment]
Let me just change it one sec and run it with the current tests.

**Comment by yurrriq** [comment]
Cool. It reduces redundancy, for sure, but _does not_ get rid of the issue with characters. I don't think there's a sensible way to do that, so we ought to just document the behaviour, suggesting to use e.g. `"-"` instead of `#\-` and give examples of both results in the docs.

**Comment by yurrriq** [comment]
Contrived examples:

``` lfe
> (clj:str "not hyphen" #\- "ated")
"not hyphen45ated"
> (clj:str "hyphen" "-" "ated")
"hyphen-ated"
```

**Comment by skovsgaard** [comment]
Okay, I'm not sure how to make this work for tests too. I can't not call `include-lib` unless I wanna be a lot of stuff short, but I also can't place it in the include file because then I'm missing `string?`. Is there a way around this?

**Comment by yurrriq** [comment]
~~I'm not sure I understand. The definition of `str` should be in `clj.lfe` so the `include-lib` call is unnecessary. I was just playing around in a shell. For the tests, you'll make a fully qualified call to `clj:str`.~~

Ah, but you can't use `fn`. Replace that with `lambda` or `clj:fn` and fully qualify the `cond->` call.

**Comment by yurrriq** [comment]
This (or similar) should work:

``` lfe
(defmacro str args
  `(lists:flatmap
    (lambda (arg)
      (clj:cond-> arg (not (clj:string? arg)) (lfe_io:print1)))
    (list ,@args)))
```

This should be fine too, if you prefer: 

``` lfe
(defmacro str args
  `(lists:flatmap
    (lambda (arg)
      (if (clj:string? arg)
        arg
        (lfe_io:print1 arg)))
    (list ,@args)))
```

**Comment by skovsgaard** [comment]
I got that to work. I wasn't thinking and just calling stuff without `clj:` prepended. Also this approach seems to treat floats a lot better than both mine and `lists:concat` do.

**Comment by yurrriq** [comment]
I've got a PR for you. Hang on.

**Comment by yurrriq** [comment]
https://github.com/skovsgaard/lfe/pull/1

**Comment by skovsgaard** [comment]
Okay, it's merged.

**Comment by yurrriq** [comment]
@rvirding, I think we're good to merge here. Note that this PR closes #269.

**Comment by skovsgaard** [comment]
Cool. I'll close the issue on this.

**Comment by skovsgaard** [comment]
Perfect. Just let me know if there's anything else missing on my end.

**Comment by yurrriq** [comment]
For the record I'm on 1.17.1 too.

**Comment by rvirding** [comment]
I'm on pandoc 1.17.0.2 but if there are any problems I will upgrade.

**Pull request merged**